URL parameter support

despeset opened the pull request:
Closes #29

despeset committed:
Adds URL parameter parsing

- Cleans the parameters out of `Request.resource`
- Creats a new Dict `Request.state[:url_params]`
- Squashes multiple instances of the same key to an array of values
- parses `foo=bar&baz=buz` to `["foo"=>"bar","baz"=>"buz"]`
- parses `foo=bar&foo=buz` to `["foo"=>["bar","buz"]]`

despeset committed:
Adds url_param convenience methods

Introduces /show/urlparams route for testing url_params.

zachallaun commented on RequestParser.jl:
```diff
@@ -106,7 +106,27 @@ on_body_cb = cfunction(on_body, HTTP_DATA_CB...)
+
+    # Handle URL variables eg. `foo/bar?a=b&c=d`
+    m = match(r"\?.*=.*", r.resource)
+    url_params = (String => Union(String,Array))[]
```
`Union(String, Array{String})` ?

zachallaun commented on RequestParser.jl:
```diff
@@ -106,7 +106,27 @@ on_body_cb = cfunction(on_body, HTTP_DATA_CB...)
+    url_params = (String => Union(String,Array))[]
+    if m != nothing
+        for set in split(split(r.resource, "?")[2], "&")
+            set = split(set, "=")
```
`key, value = split(set, "=")`

zachallaun commented on RequestParser.jl:
```diff
@@ -106,7 +106,27 @@ on_body_cb = cfunction(on_body, HTTP_DATA_CB...)
+        for set in split(split(r.resource, "?")[2], "&")
+            set = split(set, "=")
+            # Promote multiple instances of the same key to Array
+            if get(url_params, set[1], nothing) == nothing 
```
`!has(url_params, key)`

despeset committed:
Removes Array-ification of multiple keys

This would have made all devs use type checking on all URL input to ensure passing multiple keys wouldn't break their code.

despeset committed:
Adds Request.state[:raw_resource]

So squashed key value pairs can be retrieved by plucky devs

zachallaun commented:
This looks good to me now.

despeset commented:
Okay, took out the multiple key support because _some people_ ( cough cough ) think developers shouldn't have to type check every time they use a url_param.  Sheesh!  I added `Request.state[:raw_resource]` so enterprising web apps can still pull multiple values themselves.
